src/crawlers/puppeteer_crawler.js

```diff
@@ -202,292 +195,114 @@ import { validators } from '../validators';
-        }
-
-        if (options.gotoTimeoutSecs && options.gotoFunction) {
-            this.log.warning('You are using gotoTimeoutSecs with a custom gotoFunction. '
-                + 'The timeout value will not be used. With a custom gotoFunction, you need to set the timeout in the function itself.');
-        }
-
-        this.gotoFunction = gotoFunction;
-
-        this.handlePageTimeoutMillis = handlePageTimeoutSecs * 1000;
-        this.gotoTimeoutMillis = gotoTimeoutSecs * 1000;
-
-        this.puppeteerPoolOptions = {
-            ...puppeteerPoolOptions,
-            launchPuppeteerFunction,
-            launchPuppeteerOptions,
+        super({
+            ...browserCrawlerOptions,
             proxyConfiguration,
-            log: this.log,
-        };
-
-        this.puppeteerPool = null; // Constructed when .run()
-        this.persistCookiesPerSession = persistCookiesPerSession;
-        this.proxyConfiguration = proxyConfiguration;
-    }
-
-    /**
-     * Runs the crawler. Returns promise that gets resolved once all the requests got processed.
-     *
-     * @return {Promise<void>}
-     */
-    async run() {
-        if (this.isRunningPromise) return this.isRunningPromise;
-
-        await this._init();
-
-        if (this.sessionPool) {
-            this.puppeteerPoolOptions.sessionPool = this.sessionPool;
-        }
-        this.puppeteerPool = new PuppeteerPool(this.puppeteerPoolOptions);
-
-        try {
-            await super.run();
-        } finally {
-            await this.puppeteerPool.destroy();
-        }
-    }
-
-    /**
-     * Wrapper around handlePageFunction that opens and closes pages etc.
-     *
-     * @param {Object} crawlingContext
-     * @param {Request} crawlingContext.request
-     * @param {AutoscaledPool} crawlingContext.autoscaledPool
-     * @param {Session} [crawlingContext.session]
-     * @ignore
-     */
-    async _handleRequestFunction(crawlingContext) {
-        crawlingContext.page = await this.puppeteerPool.newPage();
-
-        const { page, request } = crawlingContext;
-        // eslint-disable-next-line no-underscore-dangle
-        const browserInstance = this.puppeteerPool._getBrowserInstance(page);
-        if (this.sessionPool) {
-            crawlingContext.session = browserInstance.session;
+            browserPoolOptions,
+        });
 
-            // setting cookies to page
-            if (this.persistCookiesPerSession) {
-                await page.setCookie(...crawlingContext.session.getPuppeteerCookies(request.url));
+        this.browserPool.postLaunchHooks.push(({ error, session }) => {
+            // It would be better to compare the instances,
+            // but we don't have access to puppeteer.errors here.
+            if (error && error.constructor.name === 'TimeoutError') {
+                handleRequestTimeout(session, error.message);
             }
-        }
-
-        const { session } = crawlingContext;
+        });
```

Comment:
same problem as with playwright, the error is not there

Reply:
plus it really should not be a post-launch hook, it has nothing to do with browser launches